pxf/pxf-json/src/main/java/org/apache/hawq/pxf/plugins/json/JsonResolver.java

```diff
@@ -0,0 +1,255 @@
+
+	public JsonResolver(InputData inputData) throws Exception {
+		super(inputData);
+		oneFieldList = new ArrayList<OneField>();
+		mapper = new ObjectMapper(new JsonFactory());
+
+		// Pre-generate all column structure attributes concerning the JSON to column value resolution.
+		columnDescriptorCache = new ColumnDescriptorCache[inputData.getColumns()];
+		for (int i = 0; i < inputData.getColumns(); ++i) {
+			ColumnDescriptor cd = inputData.getColumn(i);
+			columnDescriptorCache[i] = new ColumnDescriptorCache(cd);
+		}
+	}
+
+	@Override
+	public List<OneField> getFields(OneRow row) throws Exception {
+		oneFieldList.clear();
+
+		String jsonRecordAsText = row.getData().toString();
+
+		JsonNode root = decodeLineToJsonNode(jsonRecordAsText);
+
+		if (root == null) {
+			LOG.warn("Return null-fields row due to invalid JSON:" + jsonRecordAsText);
+		}
+
+		// Iterate through the column definition and fetch our JSON data
+		for (ColumnDescriptorCache column : columnDescriptorCache) {
+
+			// Get the current column description
+
+			if (root == null) {
+				// Return empty (e.g. null) filed in case of malformed json.
+				addNullField(column.getColumnType());
+			} else {
+
+				// Move down the JSON path to the final name
+				JsonNode node = getPriorJsonNode(root, column.getProjections());
+
+				// If this column is an array index, ex. "tweet.hashtags[0]"
+				if (column.isArrayName()) {
+
+					// Move to the array node
+					node = node.get(column.getArrayNodeName());
+
+					// If this node is null or missing, add a null value here
+					if (node == null || node.isMissingNode()) {
+						addNullField(column.getColumnType());
+					} else if (node.isArray()) {
+						// If the JSON node is an array, then add it to our list
+						addFieldFromJsonArray(column.getColumnType(), node, column.getArrayIndex());
+					} else {
+						throw new IllegalStateException(column.getArrayNodeName() + " is not an array node");
+					}
+				} else {
+					// This column is not an array type
+					// Move to the final node
+					node = node.get(column.getLastProjection());
+
+					// If this node is null or missing, add a null value here
+					if (node == null || node.isMissingNode()) {
+						addNullField(column.getColumnType());
+					} else {
+						// Else, add the value to the record
+						addFieldFromJsonNode(column.getColumnType(), node);
+					}
+				}
+			}
+		}
+
+		return oneFieldList;
+	}
+
+	/**
+	 * Iterates down the root node to the child JSON node defined by the projs path.
+	 * 
+	 * @param root
+	 *            node to to start the traversal from.
+	 * @param projs
+	 *            defines the path from the root to the desired child node.
+	 * @return Returns the child node defined by the root and projs path.
+	 */
+	private JsonNode getPriorJsonNode(JsonNode root, String[] projs) {
+
+		// Iterate through all the tokens to the desired JSON node
+		JsonNode node = root;
+		for (int j = 0; j < projs.length - 1; ++j) {
```

Comment:
Update: I've refactored the ColumDescriptorCache to compute a *normalized* projections. In case of array it substitutes the last path element with the name of the array. In this way the JsonResolver#getFields() is simplified. (take a look)